source/GpioInterrupt.cpp

```diff
@@ -0,0 +1,191 @@
+    }
+    return hr;
+}
+
+/// Method to attach to an interrupt on a GPIO port bit with information return.
+HRESULT GpioInterruptsClass::attachInterruptEx(ULONG pin, std::function<void(PDMAP_WAIT_INTERRUPT_NOTIFY_BUFFER)> func, ULONG mode, HANDLE hController)
+{
+    HRESULT hr = S_OK;
+    static IOControlCode^ AttachIntCode = ref new IOControlCode(FILE_DEVICE_DMAP, 0x105, IOControlAccessMode::Any, IOControlBufferingMethod::Buffered);
+    HANDLE hIntController = hController;
+    HANDLE hIntEnableEvent = m_hIntEnableEvent;
+
+    // Tell the driver to attach the interrupt.
+    if (SUCCEEDED(hr))
+    {
+        auto writer = ref new DataWriter;
+        writer->ByteOrder = ByteOrder::LittleEndian;
+        writer->WriteUInt16((uint16_t)pin);
+        writer->WriteUInt16((uint16_t)mode);
+        IBuffer^ buffer = writer->DetachBuffer();
+
+        hr = SendIOControlCodeToController(
+            hIntController,
+            AttachIntCode,
+            buffer,
+            nullptr,
+            INFINITE
+            );
+    }
+
+    if (SUCCEEDED(hr))
+    {
+        Concurrency::create_task([pin, func, mode, hIntController, hIntEnableEvent]()
+        {
+            HRESULT hr = S_OK;
+            static IOControlCode^ WaitIntCode = ref new IOControlCode(FILE_DEVICE_DMAP, 0x107, IOControlAccessMode::Any, IOControlBufferingMethod::Buffered);
+            auto writer = ref new DataWriter;
+            writer->ByteOrder = ByteOrder::LittleEndian;
+            writer->WriteUInt32(pin);
+            IBuffer^ requestBuffer = writer->DetachBuffer();
+            IBuffer^ replyBuffer = ref new Buffer(sizeof(DMAP_WAIT_INTERRUPT_NOTIFY_BUFFER));
+            BOOL continueIo = TRUE;
+
+            // Until the driver starts cancelling interrupt wait requests on this pin:
+            do
+            {
+                hr = SendIOControlCodeToController(
+                    hIntController,
+                    WaitIntCode,
+                    requestBuffer,
+                    replyBuffer,
+                    INFINITE
+                    );
+
+                if (hr == ERROR_OPERATION_ABORTED)
+                {
+                    continueIo = FALSE;
+                }
+
+                if (continueIo)
+                {
+                    // Make a byte array copy of the buffer sent back with the wait request completion.
+                    BYTE rawBuffer[sizeof(DMAP_WAIT_INTERRUPT_NOTIFY_BUFFER)];
+                    auto reader = DataReader::FromBuffer(replyBuffer);
+                    for (int i = 0; i < sizeof(DMAP_WAIT_INTERRUPT_NOTIFY_BUFFER); i++)
+                    {
+                        rawBuffer[i] = reader->ReadByte();
+                    }
+
+                    // Wait for interrupt delivery to be enabled.
+                    WaitForSingleObject(hIntEnableEvent, INFINITE);
+
+                    // Call the interrupt callback routine.
+                    func((PDMAP_WAIT_INTERRUPT_NOTIFY_BUFFER) rawBuffer);
```

Comment:
That would be a good thing to document because people should know that performing major work in their callback may affect how their app runs.  It doesn't block other interrupts, it only blocks itself.